### ISSUE TITLE
Running coq_makefile from dune

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#16307]
@Alizter your lengthy comments [here](https://github.com/coq/coq/pull/15888#discussion_r896666832) were a bit overwhelming.  My question was:

> What is your command line to `make` for the coq_makefile-generated makefile?

But I think my real question was "how do I run coq_makefile from the build environment"?

It looks like just `dune exec -- coq_makefile` builds the coq_makefile binary (if necessary) and then runs it.  What is the equivalent using `make -f Makefile.dune`?

```
$ make -f Makefile.dune coq_makefile
make: *** No rule to make target 'coq_makefile'.  Stop.
$ make -f Makefile.dune exec -- coq_makefile --version
make: *** No rule to make target 'exec'.  Stop.
$ make -f Makefile.dune coq_makefile --version           (* THIS IS NOT --version INFO FROM COQ_MAKEFILE *)
GNU Make 4.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by jfehrle** [comment]
Second question: running the `coq_makefile`-generated makefile uses the installed version of `coqc` rather than `_build/**/coqc` in the source tree.  That causes at least the version mismatch (below).  I think the easiest thing is to install coq from the dune build.  Can you tell me how to do that?  I want that to be in a fresh directory, not in opam. (I don't want to mess up the opam version.)

```
$ make -f CoqMakefile
COQDEP VFILES
Project file "_CoqProject": Syntax error: Unknown option
COQDEP VFILES
Project file "_CoqProject": Syntax error: Unknown option
W: This Makefile was generated by Coq 8.17+alpha
W: while the current Coq version is 8.15.2
COQDEP VFILES
Project file "_CoqProject": Syntax error: Unknown option
make[1]: *** No rule to make target '.CoqMakefile.d', needed by 'testcd.vo'.  Stop.
CoqMakefile:408: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by Alizter** [comment]
I apologise for my information-dense reply. Let me try to clear a few things up. 

Firstly, a quick comment: `make -f Makefile.dune` is unneeded now. Due to the work in #15560, `Makefile` has become a simple wrapper around `Makefile.dune` and soon they will be the same file. Therefore only `make` is necessary.
tldr: don't worry about separate makefiles you only have to run `make`.

>But I think my real question was "how do I run coq_makefile from the build environment"?

In my reply, I was attempting to describe how you might go about writing a shim. The reason we need a shim for running coq_makefile is because most Coq binaries need some things present in their environment. This ranges from stdlib files to certain misc. files. When you run `dune exec -- coq_makefile`, Dune will build `coq_makefile` correctly, however it does not know about the special environment we need to set up for it. This is the reason we have shims.

There are no targets in `Makefile.dune` for building `coq_makefile` today. If you want to build something, you will have to do two things:
1. Run `make dunestrap`. This will generate all the rules for .vo files in the stdlib etc.
2. Run `dune build something`.

Step 1 only needs to be done once, unless you are adding new .v files to the stdlib. When you do `make world` both steps are done automatically.

In order to build `coq_makefile` with the stdlib rules present, do 1 at least once. and then do `dune build tools/coq_makefile.exe` to build the executable or alternatively `dune exec -- coq_makefile` to build and also run it. Note that `dune exec` can find the executable and you don't need to point to its location like for `dune build.

I can write a shim for you if you wish. This would be used the same way you do `dune exec -- dev/shim/coqc-prelude`. (this is step 2).

It is really a bug that `coq_makefile` requires the presence of the stdlib so strictly today, but fixing this will require more work.

**Comment by SkySkimmer** [comment]
I just `dune build coq-core.install` and use `_build/install/default/bin/coq_makefile`

**Comment by ejgallego** [comment]
The right way to run `coq_makefile` is indeed to use `dune exec` , however, we are still missing a feature in dune that allows us to specify runtime deps, so `dune build -- coq_makefile` will just build coq_makefile, but `dune exec -- ` will also build the deps. Shims are just a workaround for that missing feature.

**Comment by jfehrle** [comment]
Thanks guys.
This worked for me, much simpler:
To build coq_makefile and run it: `dune exec -- coq_makefile -f _CoqProject -o CoqMakefile`

And this to run the generated makefile:

```
export COQBIN=_build/install/default/bin
make -f CoqMakefile
```

I verified that these called the correct coq_makefile, coqdep and coqc.
